### ISSUE TITLE
Incorrect region highlighted for "No matching constructor" error on nested classes 

### ISSUE THREAD

**Issue by vogelsgesang** [clangd/clangd#1035]
![Screenshot 2022-02-25 at 15 38 47](https://user-images.githubusercontent.com/6820896/155734050-d752a47b-be9e-4c22-bde5-b2e56e0e22ab.png)

As you can see, only `Foo` is underlined. Instead, I would expect `Foo::Bar` to be highlighted.

(clangd version 15.0.0, commit 27d39e4da0b45a, inside NeoVim)

**Comment by sam-mccall** [comment]
Yeah. This diagnostics gets emitted with the initialization location and the args range:

```
Foo::Bar{"..."};
^       ~~~~~~~
```

Clangd treats the location as primary. (It'll expand to cover the range, but only for a range containing the primary location).
I think we can ask clang to emit both ranges as well as the location, then clangd will expand the location to cover the type.

---

If the location was Bar rather than Foo we'd be at least as happy I think. SemaInit does track begin/primary/end locations for initialization location, but in this case (function cast) it's initialized from a single range (the type range) and primary is set to the beginning of the range. This is because types in general don't have a "primary" location. Here the qualifier is clearly less relevant but that doesn't generalize well.

**Comment by sam-mccall** [comment]
Ah, not so easy.

CastOperation in SemaCast keeps track of ranges for the target (DestRange) and the whole cast expression (OpRange).
But it only passes OpRange to when setting up an InitializationKind for a functional cast.

So in fact we have only this range:
```
return Foo::Bar{"incorrect type"};
       ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

And then when diagnosing, we use the start of it as the location and reconstruct the brace range for the secondary location.

**Comment by vogelsgesang** [comment]
Thanks for looking into this!

I think, if we used the range
```
return Foo::Bar{"incorrect type"};
       ~~~~~~~~~~~~~~~~~~~~~~~~~~
```
also for `clangd`, that would solve the confusion here...
It's a less targeted range, but it's less confusing at least to me.

On the other hand, why does clang highlight the complete parameter list in the first place?
Would it make sense to also change clang to report such errors with
```
return Foo::Bar{"incorrect type"};
       ~~~~~~~~
```
during normal compilation?

On a separate note: I just realized, that this bug not only impacts nested classes, but also namespaces...

![Screenshot 2022-03-01 at 13 37 50](https://user-images.githubusercontent.com/6820896/156170305-f19c9ef4-ba79-4e4c-8686-eb57568c7f48.png)
